File: app/src/main/java/com/stephenbain/obliquestrategies/view/HomePresenter.java
package com.stephenbain.obliquestrategies.view;

import com.stephenbain.obliquestrategies.util.NullObject;

import javax.inject.Inject;

public class HomePresenter {

    private static final HomeView NULL_VIEW = NullObject.create(HomeView.class);

    private HomeView view = NULL_VIEW;

    @Inject
    public HomePresenter() {

    }

    public void attach(HomeView view) {
        this.view = view;
    }

    public void detach() {
        view = NULL_VIEW;
    }

    public void present() {

    }

    interface HomeView {

    }
}

File: app/src/main/java/com/stephenbain/obliquestrategies/view/HomePresenter.kt
package com.stephenbain.obliquestrategies.view

import com.stephenbain.obliquestrategies.data.StrategyService
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

class HomePresenter @Inject constructor(private val strategyService: StrategyService) {

    private var view: HomeView? = null

    fun attach(view: HomeView) {
        this.view = view
    }

    fun detach() {
        view = null
    }

    fun present() {
        refreshStrategy()
        view?.setClickListener(::refreshStrategy)
    }

    private fun refreshStrategy() {
        strategyService.requestStrategy()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe({ cardResponse -> view?.setStrategy(cardResponse.strategy) })
    }

    interface HomeView {
        fun setStrategy(strategy: String)
        fun setClickListener(clickListener: () -> Unit)
    }

}